### PR TITLE
Lazy Logging

### PR DESCRIPTION
**Pull request by tusharmath**
## Feature
A lot of times, an effect should be executed only for logging purposes. This can cause performance issues even when the logging is turned off since the effect will still get evaluated.

## Current
```scala
for {
  f <- foo
  b <- bar
  _ <- log.debug(b) // b is needed only for logging purposes
} yield f
```

## Proposed
```scala
for {
  f <- foo
  _ <- log.debug(bar) // bar get's evaluated only when LogLevel is "debug"
} yield f
```

### CODE REVIEW

**Comment by pshemass** [comment]
I'm not sure about that. @jdegoes what do you think?

File: core/src/main/scala/zio/logging/Logger.scala [hunk 1]
```diff
@@ -23,12 +23,22 @@ trait Logger[-A] { self =>
...
   /**
    * Logs the specified element at the error level.
    */
   def error(line: => A): UIO[Unit] =
     self.log(LogLevel.Error)(line)
 
+  /**
+   * Evaluates the specified element based on the LogLevel set and logs at the error level
+   */
+  def error(line: UIO[A]): UIO[Unit] = line >>= error
```
Comment (jdegoes) [inline core/src/main/scala/zio/logging/Logger.scala:40]: This should be called `errorM`, by convention.

File: core/src/main/scala/zio/logging/Logger.scala [hunk 2]
```diff
@@ -37,6 +47,11 @@ trait Logger[-A] { self =>
       self.log(LogLevel.Error)(line)
     }
 
+  /**
+   * Evaluates the specified element based on the LogLevel set and logs at the error level
+   */
+  def error(line: UIO[A], cause: Cause[Any]): UIO[Unit] = line.flatMap(l => error(l, cause))
```
Comment (jdegoes) [inline core/src/main/scala/zio/logging/Logger.scala:53]: This should be called `errorM`, by convention.

File: core/src/main/scala/zio/logging/Logger.scala [hunk 3]
```diff
@@ -57,6 +72,11 @@ trait Logger[-A] { self =>
   def info(line: => A): UIO[Unit] =
     self.log(LogLevel.Info)(line)
 
+  /**
+   * Evaluates the specified element based on the LogLevel set and logs at the info level
+   */
+  def info(line: UIO[A]): UIO[Unit] = line >>= info
```
Comment (jdegoes) [inline core/src/main/scala/zio/logging/Logger.scala:78]: This should be called `infoM`, by convention.

File: core/src/main/scala/zio/logging/Logger.scala [hunk 4]
```diff
@@ -111,9 +131,19 @@ trait Logger[-A] { self =>
   def trace(line: => A): UIO[Unit] =
     self.log(LogLevel.Trace)(line)
 
+  /**
+   * Evaluates the specified element based on the LogLevel set and logs at the trace level
+   */
+  def trace(line: UIO[A]): UIO[Unit] = line >>= trace
```
Comment (jdegoes) [inline core/src/main/scala/zio/logging/Logger.scala:137]: This should be called `traceM`, by convention.

File: core/src/main/scala/zio/logging/Logger.scala [hunk 5]
```diff
@@ -111,9 +131,19 @@ trait Logger[-A] { self =>
...
   /**
    * Logs the specified element at the warn level.
    */
   def warn(line: => A) =
     self.log(LogLevel.Warn)(line)
+
+  /**
+   * Evaluates the specified element based on the LogLevel set and logs at the warn level
+   */
+  def warn(line: UIO[A]): UIO[Unit] = line >>= warn
```
Comment (jdegoes) [inline core/src/main/scala/zio/logging/Logger.scala:148]: This should be called `warnM`, by convention.

**Comment by jdegoes** [comment]
With slight modifications, looks great to me! I left some comments. Thanks @tusharmath!

**Comment by pshemass** [comment]
> With slight modifications, looks great to me! I left some comments. Thanks @tusharmath!

maybe I don't see something but @tusharmath proposed "Lazy Logging" - based on his comment `bar` should be evaluated only when `LogLevel.Debug` is set. This is not the case based on his code. I agree with naming convention though :)

@tusharmath could you please add test for this?

**Comment by tusharmath** [comment]
@pshemass Yes exactly. I know the PR doesn't actually do what it says. I guess, this requires a lot more deeper integration with the loggers.

**Comment by pshemass** [comment]
@tusharmath don't worry :) after all if you change those methods names as @jdegoes suggested this code is still useful and can be merged.

**Pull request merged**